common/bytes.go

```diff
@@ -0,0 +1,152 @@
+package common
+
+import (
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// Encoder is a global setting for all byte encoding
+// This is the default.  Please override in the main()/init()
+// of your program to change how byte slices are presented
+//
+// In addition to these implementation, you can also find
+// BTCEncoder and FlickrEncoder that use base58 variants in
+// github.com/tendermint/go-wire/data/base58
+var (
+	encoder       ByteEncoder = hexEncoder{}
+	HexEncoder                = hexEncoder{}
+	B64Encoder                = base64Encoder{base64.URLEncoding}
+	RawB64Encoder             = base64Encoder{base64.RawURLEncoding}
+)
+
+func SetByteEncoder(e ByteEncoder) {
+	if encoder == nil {
+		encoder = e
+	} else {
+		panic("ByteEncoder already set.")
+	}
+}
+
+func GetByteEncoder() ByteEncoder {
+	return encoder
+}
+
+// For testing.
+func clearByteEncoder() {
+	encoder = nil
+}
+
+// Bytes is a special byte slice that allows us to control the
+// serialization format per app.
+//
+// Thus, basecoin could use hex, another app base64, and a third
+// app base58...
+type Bytes []byte
+
+// Marshal needed for protobuf compatibility
+func (b Bytes) Marshal() ([]byte, error) {
+	return b, nil
+}
+
+// Unmarshal needed for protobuf compatibility
+func (b *Bytes) Unmarshal(data []byte) error {
+	*b = data
+	return nil
+}
+
+func (b Bytes) MarshalJSON() ([]byte, error) {
+	return encoder.Marshal(b)
+}
+
+func (b *Bytes) UnmarshalJSON(data []byte) error {
+	ref := (*[]byte)(b)
+	return encoder.Unmarshal(ref, data)
+}
+
+// Allow it to fulfill various interfaces in light-client, etc...
+func (b Bytes) Bytes() []byte {
+	return b
+}
+
+// String gets a simple string for printing (the json output minus quotes)
+func (b Bytes) String() string {
+	raw, err := encoder.Marshal(b)
+	l := len(raw)
+	if err != nil || l < 2 {
+		return "Bytes<?????>"
+	}
+	return string(raw[1 : l-1])
+}
+
+// ByteEncoder handles both the marshalling and unmarshalling of
+// an arbitrary byte slice.
+//
+// All Bytes use the global Encoder set in this package.
+// If you want to use this encoding for byte arrays, you can just
+// implement a simple custom marshaller for your byte array
+//
+//   type Dings [64]byte
+//
+//   func (d Dings) MarshalJSON() ([]byte, error) {
+//     return data.GetEncoder().Marshal(d[:])
+//   }
+//
+//   func (d *Dings) UnmarshalJSON(enc []byte) error {
+//     var ref []byte
+//     err := data.GetEncoder().Unmarshal(&ref, enc)
+//     copy(d[:], ref)
+//     return err
+//   }
+type ByteEncoder interface {
+	Marshal(bytes []byte) ([]byte, error)
+	Unmarshal(dst *[]byte, src []byte) error
```

Comment:
Pros: Good work on the Unmarshal! Your signature will help gut out allocations and byte slice copies everytime, by enabling byte slice reuse.
```go
var sav []byte
if err := cmn.ByteCodec.Unmarshal(&sav, data); err != nil {
}
// More stuff
if err := cmn.ByteCodec.Unmarshal(&sav, pipedIn); err != nil {
}
```
as opposed to the traditional style as from the Go stdlib which will always allocate a slice no matter what
```go
var sav []byte
blob, err := cmn.ByteCodec.Unmarshal(data)
if err != nil {
}
copy(sav, blob)
// More stuff
if err := cmn.ByteCodec.Unmarshal(&sav, pipedIn); err != nil {
}
```
However, the user doesn't need to worry about pre-allocating a slice before Unmarshaling.

Cons:
* With the commendation above though, I haven't yet seen any codec employing byte slice reuse. HexEncoder does `*dst, err = hex.DecodeString` which still performs the byte slice allocation, assigning to `*dst` and completely not reusing the preallocated slice(if that was the intention of the user).
* The new signature if employing byte slice reuse makes it so that the user has to think about how many bytes that they are expecting. If we aren't going to be resuing pre-allocated slices, perhaps let's just revert to the old style?

Perhaps in the future, let's work on fulfilling the purpose of that signature `(*[]byte, []byte)`. The approach is still commendable though.